src/compiler/program.ts

```diff
@@ -1160,6 +1206,75 @@ namespace ts {
 
         return program;
 
+        function collectCompilerExtensions(): ExtensionCollectionMap {
+            const extOptions = options.extensions;
+            const extensionNames = (extOptions instanceof Array) ? extOptions : getKeys(extOptions);
+            return groupBy(flatten(map(filter(map(extensionNames, name => {

```

Comment:
There's too much going on here, and its difficult to correlate which callback applies to which function. Consider breaking this up into multiple statements for readability.